server/src/main/java/org/elasticsearch/common/io/stream/ByteArrayStreamInput.java

```diff
@@ -84,7 +84,10 @@ protected void ensureCanReadBytes(int length) throws EOFException {
     }
 
     @Override
-    public byte readByte() {
+    public byte readByte() throws IOException {
+        if (limit - pos <= 0) {
```

Comment:
There is an existing helper, `StreamInput.ensureCanReadBytes`, which this could use? However, this seems to be getting into a broader question of validation. Do we want to pay the cost of checking these bounds on every access? If so, we should be consistent. readBytes, skipBytes, even setPosition would need similar validation. Additionally we should have checks in the unit tests for this implementation verifying this validation.

I'm not sure what the additional checks are getting us though. I understand the read() case, since that contract is via InputStream. But this method and the others I mention are with StreamInput, and the general loose contract is that the caller knows best, and we favor speed in decoding over inline validation.